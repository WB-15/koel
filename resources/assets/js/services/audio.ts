export const audio = {
  context: null as AudioContext | null,
  source: null as MediaElementAudioSourceNode | null,
  element: null as HTMLMediaElement | null,

  init (element: HTMLMediaElement): void {
    const AudioContext = window.AudioContext ||
      window.webkitAudioContext ||
      window.mozAudioContext ||
      window.oAudioContext ||
      window.msAudioContext

    this.context = new AudioContext()
    this.source = this.context.createMediaElementSource(element)
    this.element = element
  },

  getContext (): AudioContext {
    return this.context!
  },

  getSource (): MediaElementAudioSourceNode {
    return this.source!
  },

  getElement (): HTMLMediaElement {
    return this.element!
  }
}
