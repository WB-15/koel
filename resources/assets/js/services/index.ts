export * from './httpService'
export * from './downloadService'
export * from './localStorageService'
export * from './playbackService'
export * from './youTubeService'
export * from './socketService'
export * from './audioService'
export * from './uploadService'
export * from './authService'
