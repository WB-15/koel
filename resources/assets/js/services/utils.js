export default {
    /**
     * Convert a duration in seconds into H:i:s format.
     * If H is 0, it will be ommited.
     */
    secondsToHis(d) {
        d = parseInt(d);
        
        var s = d%60;

        if (s < 10) {
            s = '0' + s;
        }

        var i = Math.floor((d/60)%60);

        if (i < 10) {
            i = '0' + i;
        }

        var h = Math.floor(d/3600);

        if (h < 10) {
            h = '0' + h;
        }

        return (h === '00' ? '' : h + ':') + i + ':' + s;
    },

    /**
     * Parse the validation error from the server into a flattened array of messages.
     * 
     * @param  {Object}  error  The error object in JSON format.
     *                                    
     * @return {Array}
     */
    parseValidationError(error) {
        return Object.keys(error).reduce((messages, field) => messages.concat(error[field]), []);
    }
};
