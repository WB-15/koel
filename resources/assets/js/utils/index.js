export * from './filters'
export * from './formatters'
export * from './supports'
export * from './common'
