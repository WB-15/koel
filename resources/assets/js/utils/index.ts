export * from './event'
export * from './alerts'
export * from './filters'
export * from './formatters'
export * from './supports'
export * from './common'
export * from './$'
export * from './helpers'
export * from './fileReader'
export * from './directoryReader'
