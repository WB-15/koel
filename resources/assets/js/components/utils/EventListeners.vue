<template>
  <span></span>
</template>

<script lang="ts" setup>
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import isMobile from 'ismobilejs'
import router from '@/router'
import { authService } from '@/services'
import { playlistStore, preferenceStore, userStore } from '@/stores'
import { alerts, eventBus, forceReloadWindow } from '@/utils'

eventBus.on({
  PLAYLIST_DELETE (playlist: Playlist) {
    const destroy = async () => {
      await playlistStore.delete(playlist)
      alerts.success(`Deleted playlist "${playlist.name}."`)
      router.go('home')
    }

    if (!playlist.is_smart && !playlist.songs.length) {
      destroy()
    } else {
      alerts.confirm(`Delete the playlist "${playlist.name}"?`, destroy)
    }
  },

  /**
   * Log the current user out and reset the application state.
   */
  async LOG_OUT () {
    await userStore.logout()
    authService.destroy()
    forceReloadWindow()
  },

  'KOEL_READY': () => router.init(),

  /**
   * Hide the panel away if a main view is triggered on mobile.
   */
  'LOAD_MAIN_CONTENT': () => isMobile.phone && (preferenceStore.showExtraPanel = false)
})
</script>
