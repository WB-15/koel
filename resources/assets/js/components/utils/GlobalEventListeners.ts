/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */

import { defineComponent } from 'vue'
import { authService } from '@/services'
import { playlistFolderStore, playlistStore, userStore } from '@/stores'
import { eventBus, forceReloadWindow, requireInjection } from '@/utils'
import { RouterKey } from '@/symbols'
import { useDialogBox, useMessageToaster } from '@/composables'

export const GlobalEventListeners = defineComponent({
  setup (props, { slots }) {
    const { toastSuccess } = useMessageToaster()
    const { showConfirmDialog } = useDialogBox()
    const router = requireInjection(RouterKey)

    eventBus.on('PLAYLIST_DELETE', async playlist => {
      if (await showConfirmDialog(`Delete the playlist "${playlist.name}"?`)) {
        await playlistStore.delete(playlist)
        toastSuccess(`Playlist "${playlist.name}" deleted.`)
        router.go('home')
      }
    }).on('PLAYLIST_FOLDER_DELETE', async folder => {
      if (await showConfirmDialog(`Delete the playlist folder "${folder.name}"?`)) {
        await playlistFolderStore.delete(folder)
        toastSuccess(`Playlist folder "${folder.name}" deleted.`)
        router.go('home')
      }
    }).on('LOG_OUT', async () => {
      await userStore.logout()
      authService.destroy()
      forceReloadWindow()
    })

    return () => slots.default?.()
  }
})
