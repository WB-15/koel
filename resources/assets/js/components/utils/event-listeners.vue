<template>
  <span></span>
</template>
<script lang="ts" setup>
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import isMobile from 'ismobilejs'
import router from '@/router'
import { auth } from '@/services'
import { playlistStore, preferenceStore, userStore } from '@/stores'
import { alerts, eventBus, forceReloadWindow } from '@/utils'

eventBus.on({
  PLAYLIST_DELETE (playlist: Playlist) {
    const destroy = async () => {
      await playlistStore.delete(playlist)
      alerts.success(`Deleted playlist "${playlist.name}."`)
      router.go('home')
    }

    if (!playlist.songs.length) {
      destroy()
    } else {
      alerts.confirm(`Delete the playlist "${playlist.name}"?`, destroy)
    }
  },

  /**
   * Log the current user out and reset the application state.
   */
  async LOG_OUT () {
    await userStore.logout()
    auth.destroy()
    forceReloadWindow()
  },

  'KOEL_READY': () => router.init(),

  /**
   * Hide the panel away if a main view is triggered on mobile.
   */
  'LOAD_MAIN_CONTENT': () => isMobile.phone && (preferenceStore.showExtraPanel = false)
})
</script>
