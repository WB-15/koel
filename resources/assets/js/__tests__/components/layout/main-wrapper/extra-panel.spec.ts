import Component from '@/components/layout/main-wrapper/extra-panel.vue'
import factory from '@/__tests__/factory'
import { eventBus } from '@/utils'
import { songInfo } from '@/services'
import { mock } from '@/__tests__/__helpers__'
import { shallow, Wrapper } from '@/__tests__/adapter'

const shallowComponent = (data: object = {}): Wrapper => shallow(Component, {
  stubs: ['lyrics-pane', 'artist-info', 'album-info', 'you-tube-video-list'],
  data: () => data
})

describe('components/layout/extra-panel', () => {
  afterEach(() => {
    jest.resetModules()
    jest.clearAllMocks()
  })

  it('renders properly', () => {
    expect(shallowComponent()).toMatchSnapshot()
  })

  it('does not have a YouTube tab if not using YouTube', async () => {
    const wrapper = shallowComponent({
      sharedState: {
        useYouTube: false
      }
    })
    await wrapper.vm.$nextTick()
    expect(shallow(Component)).toMatchSnapshot()
  })

  it('has a YouTube tab if using YouTube', async () => {
    const wrapper = shallowComponent({
      sharedState: {
        useYouTube: true
      }
    })
    await wrapper.vm.$nextTick()
    expect(wrapper.has('#extraTabYouTube')).toBe(true)
  })

  it.each<[string]>([['#extraTabLyrics'], ['#extraTabAlbum'], ['#extraTabArtist']])
  ('switches to "%s" tab', selector => {
    expect(shallowComponent().click(selector).find('[aria-selected=true]').is(selector)).toBe(true)
  })

  it('fetch song info when a new song is played', () => {
    shallowComponent()
    const song = factory('song')
    const m = mock(songInfo, 'fetch', song)
    eventBus.emit('SONG_STARTED', song)
    expect(m).toHaveBeenCalledWith(song)
  })
})
