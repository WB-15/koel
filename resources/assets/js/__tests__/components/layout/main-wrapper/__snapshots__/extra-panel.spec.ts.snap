// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/layout/ExtraPanel does not have a YouTube tab if not using YouTube 1`] = `
<section id="extra" data-testid="extra-panel" class="text-secondary showing">
  <div class="tabs">
    <div role="tablist" class="clear"><button aria-selected="true" aria-controls="extraPanelLyrics" id="extraTabLyrics" role="tab">
        Lyrics
      </button> <button aria-controls="extraPanelArtist" id="extraTabArtist" role="tab">
        Artist
      </button> <button aria-controls="extraPanelAlbum" id="extraTabAlbum" role="tab">
        Album
      </button>
      <!---->
    </div>
    <div class="panes">
      <div aria-labelledby="extraTabLyrics" id="extraPanelLyrics" role="tabpanel" tabindex="0">
        <lyricspane-stub></lyricspane-stub>
      </div>
      <div aria-labelledby="extraTabArtist" id="extraPanelArtist" role="tabpanel" tabindex="0" style="display: none;">
        <!---->
      </div>
      <div aria-labelledby="extraTabAlbum" id="extraPanelAlbum" role="tabpanel" tabindex="0" style="display: none;">
        <!---->
      </div>
      <div aria-labelledby="extraTabYouTube" id="extraPanelYouTube" role="tabpanel" tabindex="0" style="display: none;">
        <!---->
      </div>
    </div>
  </div>
</section>
`;

exports[`components/layout/ExtraPanel renders properly 1`] = `
<section id="extra" data-testid="extra-panel" class="text-secondary showing">
  <div class="tabs">
    <div role="tablist" class="clear"><button aria-selected="true" aria-controls="extraPanelLyrics" id="extraTabLyrics" role="tab">
        Lyrics
      </button> <button aria-controls="extraPanelArtist" id="extraTabArtist" role="tab">
        Artist
      </button> <button aria-controls="extraPanelAlbum" id="extraTabAlbum" role="tab">
        Album
      </button>
      <!---->
    </div>
    <div class="panes">
      <div aria-labelledby="extraTabLyrics" id="extraPanelLyrics" role="tabpanel" tabindex="0">
        <lyrics-pane-stub></lyrics-pane-stub>
      </div>
      <div aria-labelledby="extraTabArtist" id="extraPanelArtist" role="tabpanel" tabindex="0" style="display: none;">
        <!---->
      </div>
      <div aria-labelledby="extraTabAlbum" id="extraPanelAlbum" role="tabpanel" tabindex="0" style="display: none;">
        <!---->
      </div>
      <div aria-labelledby="extraTabYouTube" id="extraPanelYouTube" role="tabpanel" tabindex="0" style="display: none;">
        <!---->
      </div>
    </div>
  </div>
</section>
`;
